beacon_node/beacon_chain/src/validator_custody.rs

```diff
@@ -0,0 +1,433 @@
+        if self.persisted_is_supernode {
+            return spec.number_of_custody_groups;
+        }
+        let advertised_validator_custody = self
+            .advertised_validator_custody_count
+            .load(Ordering::Relaxed);
+
+        // If there are no validators, return the minimum custody_requirement
+        if advertised_validator_custody > 0 {
+            advertised_validator_custody
+        } else {
+            spec.custody_requirement
+        }
+    }
+
+    /// The custody count that we use to custody columns currently.
+    ///
+    /// This function should be called when figuring out how many columns we
+    /// need to custody when receiving blocks over gossip/rpc or during sync.
+    pub fn custody_group_count(&self, spec: &ChainSpec) -> u64 {
+        if self.current_is_supernode {
+            return spec.number_of_custody_groups;
+        }
+        let validator_custody_count_at_head =
+            self.validator_custody_count_at_head.load(Ordering::Relaxed);
+
+        // If there are no validators, return the minimum custody_requirement
+        if validator_custody_count_at_head > 0 {
+            validator_custody_count_at_head
+        } else {
+            spec.custody_requirement
+        }
+    }
+
+    /// Returns the count of custody columns this node must sample for block import.
+    pub fn sampling_count(&self, spec: &ChainSpec) -> u64 {
+        // This only panics if the chain spec contains invalid values
+        spec.sampling_size(self.custody_group_count(spec))
+            .expect("should compute node sampling size from valid chain spec")
+    }
+}
+
+// write a service that emits events when internal values change
+#[derive(Debug, Clone)]
+pub enum CustodyContextMessage {
+    /// The custody count changed because of a change in the
+    /// number of validators being managed.
+    ///
+    /// This should trigger actions downstream like
+    /// subscribing/unsubscribing new subnets/
+    /// backfilling required columns.
+    HeadCustodyCountChanged { new_custody_count: u64 },
+    /// The advertised custody count has changed.
+    ///
+    /// This should trigger downstream actions like setting
+    /// a new cgc value in the enr and metadata fields and
+    /// performing any related cleanup actions.
+    AdvertisedCustodyCountChanged { new_custody_count: u64 },
+}
+
+/// The custody information that gets persisted across runs.
+#[derive(Debug, Encode, Decode, Clone)]
+pub struct CustodyContextSsz {
+    advertised_validator_custody: u64,
+    validator_custody_at_head: u64,
+    persisted_is_supernode: bool,
+}
+
+impl From<&CustodyContext> for CustodyContextSsz {
+    fn from(context: &CustodyContext) -> Self {
+        CustodyContextSsz {
+            advertised_validator_custody: context
+                .advertised_validator_custody_count
+                .load(Ordering::Relaxed)
+                .clone(),
```

Comment:
these clones after `.load` are not required as `AtomicU64::load() ` returns `u64` ?

Reply:
Addressed in https://github.com/sigp/lighthouse/pull/7578/commits/1adcc4a5406911fa8677a51507a787a9c1565e2c